### PR TITLE
set error flag if http connection is closed

### LINKED ISSUE #134: set the error flag when the client closed a http connection
Currently, the fault flag of a segment is set when the client closed a http connection.
But it is actually not fatal error, because the users don't get any error response.
So the error flag should be set when the client closed a http connection.

See https://songmu.jp/riji/entry/2020-12-16-go-http499.html (written in Japanese)

### PR DESCRIPTION
**Pull request by furusax0621**
Fix #134

### CODE REVIEW

File: xrayhttp/handler.go [hunk 1]
```diff
@@ -109,6 +109,15 @@ func (tracer *httpTracer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	}
 	seg.SetHTTPResponse(responseInfo)
+
+	// Set error flag if http connection is already closed by client.
+	select {
+	case <-ctx.Done():
+		seg.SetError()
+		return
+	default:
+	}
```
Comment (shogo82148) [inline xrayhttp/handler.go:119]: 👍🏻

File: xrayhttp/handler_test.go [hunk 2]
```diff
@@ -124,6 +125,93 @@ func TestHandler(t *testing.T) {
...
+		// check optional interface of http.ResponseWriter
+		if _, ok := w.(http.Hijacker); ok {
+			t.Error("want not implement http.Hijacker, but it does")
+		}
+		if _, ok := w.(http.Flusher); !ok {
+			t.Error("want implement http.Flusher, but it doesn't")
+		}
+		if _, ok := w.(http.Pusher); ok {
+			t.Error("want not implement http.Pusher, but it does")
+		}
```
Comment (shogo82148) [inline xrayhttp/handler_test.go:142]: They are tested by `TestHandler`, so we don't need them here.

File: xrayhttp/handler_test.go [hunk 3]
```diff
@@ -124,6 +125,93 @@ func TestHandler(t *testing.T) {
...
+			w.WriteHeader(http.StatusInternalServerError)
+			if _, err := w.Write([]byte("canceled")); err != nil {
+				panic(err)
+			}
+			return
+		default:
+		}
+
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(http.StatusOK)
```
Comment (shogo82148) [inline xrayhttp/handler_test.go:157]: we know that the context is canceled.
so the handler can always return `http.StatusInternalServerError`.

File: xrayhttp/handler_test.go [hunk 4]
```diff
@@ -124,6 +125,93 @@ func TestHandler(t *testing.T) {
...
+		select {
+		case <-r.Context().Done():
+			w.Header().Set("Content-Type", "text/plain")
+			w.WriteHeader(http.StatusInternalServerError)
+			if _, err := w.Write([]byte("canceled")); err != nil {
+				panic(err)
+			}
+			return
+		default:
+		}
```
Comment (shogo82148) [inline xrayhttp/handler_test.go:154]: make the test failure directly.

```suggestion
		select {
		case <-r.Context().Done():
		default:
			t.Error("the context should be canceled, but not")
		}
```

it keeps the test simple.

**Review by shogo82148 — CHANGES_REQUESTED** [review]
the implementation looks good, but please fix the test.

**Comment by furusax0621** [comment]
@shogo82148 I modified the test according to your comments. Please review again.

**Review by shogo82148 — APPROVED** [review]
LGTM

**Pull request merged**